### ISSUE TITLE
RSSI not implemented for Node-BLE, get from Tile Directly?

### ISSUE THREAD

**Issue by NikeLaosClericus** [lesleyxyz/node-tile#3]
https://github.com/lesleyxyz/node-tile/blob/1b3bdf111df48029190b81332325dfa3c952429d/src/services/TileServiceBle.ts#L27
The code comment above lists it as a todo, but testing directly with node-ble seems to suggest that the issue lies with device.getRSSI() itself.

Alternatively, it may be possible to get a reported RSSI value directly from the Tile, per this documentation:
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsdk_nrf5_v16.0.0%2Fgroup___t_o_a___c_o_m_p_a_t_i_b_i_l_i_t_y.html&resultof=%22%74%69%6c%65%22%20%22%72%73%73%69%22%20

Listed under TOA Supported Features in Byte 2 is: 
`
bit 5 | TRM : Tile RSSI Monitoring
`

And I see hints to TOA_RSP_TRM in:
https://github.com/lesleyxyz/node-tile/blob/1b3bdf111df48029190b81332325dfa3c952429d/src/models/ToaTransaction.ts#L46
and
https://github.com/lesleyxyz/node-tile/blob/1b3bdf111df48029190b81332325dfa3c952429d/research/frida/blemon.js#L77

Is it possible to pull a Tile's RSSI using this?

**Comment by lesleyxyz** [comment]
I didn't implement this for node-ble because of the following error:
https://github.com/chrvadala/node-ble/issues/6
This issue got closed incorrectly.

If noble is able to get rssi, so should node-ble.
I'll create a new issue on node-ble later today or this weekend.

I have never seen that documentation before, it would've helped me so much to make this haha, thanks for the link.
This is indeed another option, I'll take a look later today or this weekend.

**Comment by NikeLaosClericus** [comment]
Following that node-ble issue, I did some testing.
buildog notes that "RSSI/TXPower is available until connection," but I don't think that's quite it.
I think that what's really happening is that a device connection somehow "suspends" rssi updates.

I was able to use device.getRSSI by making sure that _Discovery is running_ at the time of the call (ie, don't call adapter.stopDiscovery() on line 23 of example.js)

I threw this together for testing, and it works great, streaming rssi values that update as I move my Tile.
```
async function getRSSILoop(device){
	while(true){
                try {
                    const rssi = await device.getRSSI();
                    console.log(Date.now(), "rssi", rssi);
                } catch (error) {
                    console.log(Date.now(), "no rssi");
                }
		await sleep(1000);
	}
}
```
And I am able to connect to the device as well...
The only weird thing is that it seems like connecting "too soon" breaks getRSSI from the get go:
```
type: 'org.freedesktop.DBus.Error.InvalidArgs',
  text: "No such property 'RSSI'",
  reply: Message {
    type: 3,
    _sent: false,
    _serial: 43895,
    path: undefined,
    interface: undefined,
    member: undefined,
    errorName: 'org.freedesktop.DBus.Error.InvalidArgs',
    replySerial: 129,
    destination: ':1.4764',
    sender: ':1.9',
    signature: 's',
    body: [ "No such property 'RSSI'" ],
    flags: 1
  }
```
It's a little random, but it seems like whenever device connection happens quickly, rssi fails.
But if device connection is a little slow, slow enough for a few rssi messages to be returned, then it's fine?
but then the rssi values never get updated, so I assume that error is because there was never an rssi value to rely on.

I also tried disconnecting after, and that does seem to restore rssi.

To be honest, I don't really know how BLE works, so I don't know if this is intended behavior or not, but it seems like I just can't use getRSSI while connected/connecting.

**Comment by lesleyxyz** [comment]
Well for some reason rssi works fine with noble...
Anyway I implemented TRM (Tile RSSI Monitoring) in 1.1.1 which is now live.
This can be used with both noble & node-ble

I applied the same rssi normalization technique as they used in the app, except substracting the `txPowerOffset` which at the moment I can't figure out how to get. But I find it accurate enough.

To get RSSI through TRM you need to listen to the `tileRssi` event instead of the `rssi` event like this:
```js
service.on("tileRssi", rssi => console.log("tileRssi", rssi))
```

Only downside of TRM is that you require an authenticated session for it to work, while the normal bluetooth RSSI works without connecting or authentication.

Let me know if this solution suffices your use case

**Comment by NikeLaosClericus** [comment]
That's great! getting from tile works!
